File: leetcode/c++/LCR_193_er_cha_sou_suo_shu_de_zui_jin_gong_gong_zu_xian_lcof.cpp
// https://leetcode.cn/problems/er-cha-sou-suo-shu-de-zui-jin-gong-gong-zu-xian-lcof/description

#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
      int val;
      TreeNode *left;
      TreeNode *right;
      TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        while (root != NULL) {
            if (root->val > p->val && root->val > q->val)
                root = root->left;
            else if (root->val < p->val && root->val < q->val)
                root = root->right;
            else 
                break;
        }
        return root;
    }
};

// 迭代，时间复杂度为最差O(n)，最好O(logn)，空间复杂度为O(1)

File: leetcode/golang/LCR_193_er_cha_sou_suo_shu_de_zui_jin_gong_gong_zu_xian_lcof.go
type TreeNode struct {
         Val   int
         Left  *TreeNode
         Right *TreeNode
}

   
func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
    if root == nil {
        return nil 
    }
    if root.Val < p.Val && root.Val < q.Val {
        return lowestCommonAncestor(root.Right, p, q)
    }
    if root.Val > p.Val && root.Val > q.Val {
        return lowestCommonAncestor(root.Left, p, q)
    }
    return root
}